// @amzn/harmony-build-scripts/dist/utils/cleaner.js
/*
This code was ported over from the Harmony CLI. It's responsible for "cleaning" references to problematic Harmony URLs in
the built assets and removing the original "dev reload" script we shipped out of the box. I wish this code didn't have
to exist, but for backwards compatibility sake, I'm afraid it has to.

https://code.amazon.com/packages/HarmonyCLI/blobs/b719e89336c00e7d988c07056748e77f8413a9ec/--/src/build/harmony-app-cleaner.js
*/
import * as fs from "node:fs";
import { readdir } from "node:fs/promises";
import * as cheerio from "cheerio";
// Just to have the output of this command stand out in the build log
function formatCleaningMessage(filename, message) {
    return `  - ${filename}: ${message}`;
}
function log(filename, message) {
    console.log(formatCleaningMessage(filename, message));
}
function warn(filename, message) {
    console.warn(formatCleaningMessage(filename, `WARNING: ${message}`));
}
// starts with https://<stage>.console.harmony.a2z.com
const HARMONY_CONSOLE_ABSOLUTE_PATH_REGEX = /^(https?:)?\/\/([a-z]*\.)?console.harmony.a2z.com(.*)$/;
// does not start with 'http:', 'https:', 'www.', '/', '../', './'
// although ../ and ./ are in fact relative paths, we don't want to replace their
// path values since they are usually configured relative to the root application /app-name/.
const APP_RELATIVE_PATH_REGEX = /^(?!((http:)|(https:)|(www\.)|(\/)|(\.(\.)?\/))).+/;
const removeAbsoluteHarmonyPathsInFile = (directory, filename, dryRun) => {
    const data = fs.readFileSync(`${directory}/${filename}`, "utf8");
    const $ = cheerio.load(data);
    $("script").map(function (index, el) {
        const src = $(el).attr("src");
        if (src) {
            const grpMatch = src.match(HARMONY_CONSOLE_ABSOLUTE_PATH_REGEX);
            if (grpMatch && grpMatch.length) {
                if (dryRun) {
                    warn(filename, `Found src=${src} in <script> but leaving it in because --apply-legacy-cleaning wasn't specified.`);
                    return;
                }
                log(filename, `Replacing src=${src} in <script> with ${grpMatch[grpMatch.length - 1]}`);
                $(el).attr("src", grpMatch[grpMatch.length - 1]);
            }
        }
    });
    $("link").map(function (index, el) {
        const href = $(el).attr("href");
        if (href) {
            const grpMatch = href.match(HARMONY_CONSOLE_ABSOLUTE_PATH_REGEX);
            if (grpMatch && grpMatch.length) {
                if (dryRun) {
                    warn(filename, `Found href=${href} in <link>of but leaving it in because --apply-legacy-cleaning wasn't specified.`);
                    return;
                }
                log(filename, `Replacing href=${href} in <link> with ${grpMatch[grpMatch.length - 1]}`);
                $(el).attr("href", grpMatch[grpMatch.length - 1]);
            }
        }
    });
    if (!dryRun) {
        fs.writeFileSync(`${directory}/${filename}`, $.html());
    }
};
const prependAppNameToRelativePathsInFile = (directory, filename, appName, dryRun) => {
    const data = fs.readFileSync(`${directory}/${filename}`, "utf8");
    const $ = cheerio.load(data);
    // Trying to catch all the tags that use the src attribute. We're relying on the fact that this code has
    // been part of the build process from the beginning of Harmony. Please don't add or remove any tags here.
    // https://www.w3schools.com/tags/att_src.asp
    $("script, img, input, source, track, video, iframe").map(function (index, el) {
        const src = $(el).attr("src");
        if (src) {
            const grpMatch = src.match(APP_RELATIVE_PATH_REGEX);
            if (grpMatch && grpMatch.length) {
                if (dryRun) {
                    warn(filename, `Found potentially problematic relative link src=${src} in <${el.type}> but leaving it in because --apply-legacy-cleaning wasn't specified.`);
                    return;
                }
                log(filename, `Replacing src=${src} in <${el.type}> with /${appName}/${src}`);
                $(el).attr("src", `/${appName}/${src}`);
            }
        }
    });
    $("link").map(function (index, el) {
        const href = $(el).attr("href");
        if (href) {
            const grpMatch = href.match(APP_RELATIVE_PATH_REGEX);
            if (grpMatch && grpMatch.length) {
                if (dryRun) {
                    warn(filename, `Found potentially problematic relative link href=${href} in <link> but leaving it in because --apply-legacy-cleaning wasn't specified.`);
                    return;
                }
                log(filename, `Replacing href=${href} in <link> with /${appName}/${href}`);
                $(el).attr("href", `/${appName}/${href}`);
            }
        }
    });
    if (!dryRun) {
        fs.writeFileSync(`${directory}/${filename}`, $.html());
    }
};
const removeReloadScriptsInFile = (directory, filename, dryRun) => {
    const data = fs.readFileSync(`${directory}/${filename}`, "utf8");
    const $ = cheerio.load(data);
    const reloadDevScript = $("head script#reloadDevScript");
    if (reloadDevScript.length) {
        if (!dryRun) {
            log(filename, "Removing reloadDevScript");
            reloadDevScript.remove();
            fs.writeFileSync(`${directory}/${filename}`, $.html());
        }
        else {
            warn(filename, `Found script#reloadDevScript tag in <head> but leaving it in because --apply-legacy-cleaning wasn't specified.`);
        }
    }
};
/**
 * As of writing, our templates ship out of the box with references to the beta Harmony nav in order for the development
 * experience to mimic that of the app once deployed. So we need to transform
 *
 *  <script src="https://beta.console.harmony.a2z.com/navigation/harmony-navbar.js"></script>
 *
 *     into
 *
 *  <script src="/navigation/harmony-navbar.js"></script>
 *
 * so that the app will use the nav from its own stage.
 *
 * @param dir the app build directory, containing the HTML files to modify
 * @param dryRun true to indicate that no changes should be made to the files
 */
export const removeAbsoluteHarmonyPaths = async (dir, dryRun) => {
    console.log("🎸 Checking for explicit references to Harmony stages...");
    const files = await readdir(dir, { recursive: true });
    files
        .filter((filename) => filename.endsWith(".html"))
        .forEach((filename) => removeAbsoluteHarmonyPathsInFile(dir, filename, dryRun));
};
/**
 * Account for relative paths that would break once deployed, and prepend them with the name of the app.
 * For example,
 *
 *   <img src="images/foo.png" />
 *
 * might work fine during development since (at least historically) the dev server serves assets from the root.
 * But once deployed, visiting the app at console.harmony.a2z.com/my-app would try to load that image from
 * console.harmony.a2z.com/foo.png which would break.
 *
 * @param dir the app build directory, containing the HTML files to modify
 * @param appName the name of the app
 * @param dryRun true to indicate that no changes should be made to the files
 */
export const prependAppNameToRelativePaths = async (dir, appName, dryRun) => {
    console.log("🎸 Checking for problematic relative links...");
    const files = await readdir(dir, { recursive: true });
    files
        .filter((filename) => filename.endsWith(".html"))
        .forEach((filename) => prependAppNameToRelativePathsInFile(dir, filename, appName, dryRun));
};
/**
 * Removes the dev reload script that we shipped with some of the templates.
 *
 * @param directory the directory in which to find the file
 * @param dryRun true to indicate that no changes should be made to the files
 */
export const removeReloadScripts = async (dir, dryRun) => {
    console.log("🎸 Checking for development reload scripts...");
    const files = await readdir(dir, { recursive: true });
    files
        .filter((filename) => filename.endsWith(".html"))
        .forEach((filename) => removeReloadScriptsInFile(dir, filename, dryRun));
};
