// @amzn/harmony-build-tools/dist/display-devspaces-link-CMkINd7Y.js
import { resolve as c } from "path";
import { readFileSync as p } from "fs";
import { loadEnv as i } from "vite";
function v(a = "./.harmony/harmony-metadata.json") {
  return {
    name: "set-dev-cookie",
    configureServer(o) {
      const n = c(process.cwd(), a);
      let r = "";
      try {
        r = JSON.parse(p(n, "utf-8")).appName || r;
      } catch (e) {
        console.warn(`Could not read ${a}:`, e.message);
      }
      o.middlewares.use((e, t, s) => {
        t.setHeader("Set-Cookie", `Harmony-App-Override=${r}; Path=/`), s();
      });
    }
  };
}
function u(a = "./.harmony/harmony-metadata.json") {
  return {
    name: "import-nav",
    transformIndexHtml(o, n) {
      if (n.bundle)
        return o.replace("</head>", '<script src="/navigation/harmony-navbar.js"><\/script></head>');
      const r = c(process.cwd(), a);
      let e, t = "https://beta.console.harmony.a2z.com/navigation/harmony-navbar.js";
      try {
        e = JSON.parse(p(r, "utf-8")).spaceName || e;
      } catch (s) {
        console.warn(`Could not read ${a}:`, s.message);
      }
      return o.replace("</head>", `<script src="${t}"><\/script></head>`);
    }
  };
}
function f() {
  let a, o;
  return {
    name: "display-devspaces-link",
    config(n, { mode: r }) {
      const e = i(r, process.cwd(), "");
      a = e.DEVSPACE_ID, o = e.AWS_REGION;
    },
    configureServer(n) {
      const r = n.printUrls;
      n.printUrls = () => {
        if (r(), a && o) {
          const e = n.config.server.port, t = `https://${a}--${e}.${o}.prod.proxy.devspaces.amazon.dev`;
          console.log(`  \x1B[32m➜\x1B[0m  \x1B[1mDevSpace:\x1B[0m \x1B[36m${t}\x1B[0m`);
        }
      };
    }
  };
}
export {
  f as d,
  u as i,
  v as s
};
